esp32/machine_pin: Add mode and pull in machine_pin_print() as repr() function.

IhorNehrutsa opened the pull request:
**Summary**
This PR adds `mode`, `pull`, `drive` parameters to the Pin repr function.

The main advantage:
This PR allows to serialize Pin object to str and restore the Pin object from the string:
```
from machine import Pin

pin1 = Pin(27, mode=Pin.IN, pull=Pin.PULL_DOWN)
print(pin1)
pin2 = eval(str(pin1))
print(pin2)
```
Output is:
```
Pin(27, mode=Pin.IN, pull=Pin.PULL_DOWN)
Pin(27, mode=Pin.IN, pull=Pin.PULL_DOWN)
```

**Testing**
Test code is:
```
from time import sleep
from machine import Pin

PIN = 27

pin = Pin(PIN, mode=Pin.IN, pull=None)
sleep(2)
print(pin, pin())

pin = Pin(PIN, mode=Pin.IN, pull=Pin.PULL_UP)
sleep(2)
print(pin, pin())
assert pin() == 1

pin = Pin(PIN, mode=Pin.IN, pull=Pin.PULL_DOWN)
sleep(2)
print(pin, pin())
assert pin() == 0

pin = Pin(PIN, mode=Pin.OUT, value=1)
sleep(2)
print(pin, pin())
assert pin() == 1

pin = Pin(PIN, mode=Pin.OUT, value=0, drive=Pin.DRIVE_1)
sleep(2)
print(pin, pin())
assert pin() == 0
```
Output is:
```
Pin(27, mode=Pin.IN) 0
Pin(27, mode=Pin.IN, pull=Pin.PULL_UP) 1
Pin(27, mode=Pin.IN, pull=Pin.PULL_DOWN) 0
Pin(27, mode=Pin.OUT, pull=Pin.PULL_DOWN) 1
Pin(27, mode=Pin.OUT, pull=Pin.PULL_DOWN, drive=Pin.DRIVE_1) 0
```

IhorNehrutsa commented:
@jimmo 
Could you give your opinion about this PR?

jimmo commented:
Thanks @IhorNehrutsa -- I can definitely see that adding this information to the print output is useful for debugging, and making the pin serialisable is certainly a nice bonus.

However, there are two main concerns:

1. We don't ensure that the pins can be eval'ed on other ports. For example, on rp2040:

```
>>> machine.Pin.board.D12
Pin(GPIO4, mode=ALT, pull=PULL_DOWN, alt=31)
```

It's quite useful (for debugging) that it prints the mode strings rather than the numbers too. (I guess you could in theory make this work with eval with the right combination of globals passed to eval).

2. It's not a trivial decision to dedicate ~90 bytes (2 uint8_t * ~45 pins) of RAM just to tracking this state. 90 bytes isn't much, but I don't know if it justifies this? I was quite surprised to find that the IDF doesn't provide a way to extract this information from a pin, but that definitely seems to be the case in both the driver and HAL.

What if we turned `machine_pin_obj_cfg_t` into a bitfield (3 bits for gpio_mode_t, two bits for gpio_pull_mode_t). Then it would be only 45 bytes?

IhorNehrutsa commented:
I opened the ESP-IDF issue:
[GPIO feature: Add gpio_get_direction() and gpio_get_pull_mode() (IDFGH-10987) #12176](https://github.com/espressif/esp-idf/issues/12176)

IhorNehrutsa commented:
Function [gpio_dump_io_configuration()](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/gpio.html#_CPPv426gpio_dump_io_configurationP4FILE8uint64_t) 
Dump IO configuration information to console.

Commit https://github.com/espressif/esp-idf/commit/321f628ff52fb91e148c2a2493e5d8c2b18d02a1
feat(gpio): add a dump API to dump IO configurations

IhorNehrutsa commented:
```
        bool pu, pd, ie, oe, od, slp_sel;
        uint32_t drv, fun_sel, sig_out;
        gpio_hal_get_io_config(gpio_context.gpio_hal, gpio_num, &pu, &pd, &ie, &oe, &od, &drv, &fun_sel, &sig_out, &slp_sel);
```

IhorNehrutsa commented:
feat(gpio): add a dump API to dump IO configurations (IDFGH-11366)
https://github.com/espressif/esp-idf/pull/12511
Cherry pick from master to release/v5.1

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.

dpgeorge commented:
Now that esp32 port uses IDF 5.2.2, are the required IDF functions available to retrieve the GPIO state?

Also, would be nice to print the string of the mode/pull, rather than just the integer (like the rp2 port does).

IhorNehrutsa committed:
esp32/machine_pin: Add mode, pull and drive to machine_pin_print().

This commit adds mode, pull, drive parameters to the Pin repr function.
This allows to serialize Pin object to str and restore the Pin object from
the string.

Signed-off-by: IhorNehrutsa <Ihor.Nehrutsa@gmail.com>

IhorNehrutsa commented:
Test code is:
```
from time import sleep
from machine import Pin

PIN = 27 # 5 #

pin = Pin(PIN, mode=Pin.IN, pull=Pin.PULL_DOWN)
sleep(2)
print(pin, pin())

pin = Pin(PIN, mode=Pin.IN, pull=Pin.PULL_UP)
sleep(2)
print(pin, pin())

pin = Pin(PIN, mode=Pin.IN, pull=None)
sleep(2)
print(pin, pin())

pin = Pin(PIN, mode=Pin.OUT, value=0)
sleep(2)
print(pin, pin())

pin = Pin(PIN, mode=Pin.OUT, value=1, drive=Pin.DRIVE_1)
sleep(2)
print(pin, pin())
```
The output is:
```
Pin(27, mode=Pin.IN, pull=Pin.PULL_DOWN) 0
Pin(27, mode=Pin.IN, pull=Pin.PULL_UP) 0
Pin(27, mode=Pin.IN) 0
Pin(27, mode=Pin.OUT) 0
Pin(27, mode=Pin.OUT, drive=Pin.DRIVE_1) 1
```

This PR requires in esp-idf
[GPIO: Add gpio_get_io_config() and xxx_is_enabled(). (IDFGH-14117)](https://github.com/espressif/esp-idf/pull/14923)

IhorNehrutsa commented:
> Now that esp32 port uses IDF 5.2.2, are the required IDF functions available to retrieve the GPIO state?
> 
> Also, would be nice to print the string of the mode/pull, rather than just the integer (like the rp2 port does).

This PR requires:
esp-idf commit https://github.com/espressif/esp-idf/commit/4535c272248d2e554ae22387fbb92d8eeb6d1e55
**esp32: IDF v.5.5 compile.** #16565

projectgus commented:
Thanks @IhorNehrutsa for persisting with this PR.

I've marked it as a draft for now, because we won't be able to review and merge it until there's a release version of ESP-IDF v5.5 that MicroPython supports. When this happens, please feel free to rebase and click "Ready for review".

(One thing to note, it will be a *very* long time before we drop support for all older ESP-IDF, so this PR will probably need some `#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 5, 0)` guards around it.)

IhorNehrutsa commented:
> #if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 5, 0) guards around it.

Done.

IhorNehrutsa commented:
Ready with IDF v.5.5

IhorNehrutsa commented:
@dpgeorge 
> Now that esp32 port uses IDF 5.2.2, are the required IDF functions available to retrieve the GPIO state?
> 
> Also, would be nice to print the string of the mode/pull, rather than just the integer (like the rp2 port does).

Ready

dpgeorge commented on ports/esp32/machine_pin.c:
```diff
@@ -129,7 +129,29 @@ gpio_num_t machine_pin_get_id(mp_obj_t pin_in) {
+        mp_printf(print, ", mode=Pin.OPEN_DRAIN");
+    } else if (gpio_io_config.ie) {
+        mp_printf(print, ", mode=Pin.IN");
+    }
```
I think the code could be made simpler doing something like:
```c
    if (gpio_io_config.oe) {
        mode = MP_QSTR_OUT;
    } else if (gpio_io_config.od) {
        mode = MP_QSTR_OPEN_DRAIN;
    } else if (gpio_io_config.ie) {
        mode = MP_QSTR_IN;
    }
    mp_printf(print, ", mode=%q.%q", MP_QSTR_Pin, mode);
```
and similarly below with the pull up/down.

IhorNehrutsa replied:
Ready

projectgus reviewed (approved):
LGTM, thanks @IhorNehrutsa!

dpgeorge commented:
Merged!  Thanks for keeping this going.
